pkg/sdn/plugin/ovscontroller.go

```diff
@@ -0,0 +1,229 @@
+	otx.AddFlow("table=111, priority=0, actions=drop")
+
+	// Table 120: multicast delivery to local pods (either from VXLAN or local pods); updated by updateLocalMulticastFlows() in pod.go
+	// eg, "table=120, priority=100, reg0=${tenant_id}, actions=output:${ovs_port_1},output:${ovs_port_2}"
+	otx.AddFlow("table=120, priority=0, actions=drop")
+
+	// Table 253: rule version note
+	otx.AddFlow("table=%d, actions=%s", VERSION_TABLE, versionNote)
+
+	err = otx.EndTransaction()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (oc *ovsController) NewTransaction() ovs.Transaction {
+	return oc.ovs.NewTransaction()
+}
+
+func (oc *ovsController) AddHostSubnetRules(subnet *osapi.HostSubnet) error {
+	otx := oc.ovs.NewTransaction()
+
+	otx.AddFlow("table=10, priority=100, tun_src=%s, actions=goto_table:30", subnet.HostIP)
+	if vnid, ok := subnet.Annotations[osapi.FixedVNIDHostAnnotation]; ok {
+		otx.AddFlow("table=50, priority=100, arp, nw_dst=%s, actions=load:%s->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,output:1", subnet.Subnet, vnid, subnet.HostIP)
+		otx.AddFlow("table=90, priority=100, ip, nw_dst=%s, actions=load:%s->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,output:1", subnet.Subnet, vnid, subnet.HostIP)
+	} else {
+		otx.AddFlow("table=50, priority=100, arp, nw_dst=%s, actions=move:NXM_NX_REG0[]->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,output:1", subnet.Subnet, subnet.HostIP)
+		otx.AddFlow("table=90, priority=100, ip, nw_dst=%s, actions=move:NXM_NX_REG0[]->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,output:1", subnet.Subnet, subnet.HostIP)
+	}
+
+	return otx.EndTransaction()
+}
+
+func (oc *ovsController) DeleteHostSubnetRules(subnet *osapi.HostSubnet) error {
+	otx := oc.ovs.NewTransaction()
+	otx.DeleteFlows("table=10, tun_src=%s", subnet.HostIP)
+	otx.DeleteFlows("table=50, arp, nw_dst=%s", subnet.Subnet)
+	otx.DeleteFlows("table=90, ip, nw_dst=%s", subnet.Subnet)
+	return otx.EndTransaction()
+}
+
+func (oc *ovsController) AddServiceRules(service *kapi.Service, netID uint32) error {
+	otx := oc.ovs.NewTransaction()
+
+	action := fmt.Sprintf(", priority=100, actions=load:%d->NXM_NX_REG1[], load:2->NXM_NX_REG2[], goto_table:80", netID)
+
+	// Add blanket rule allowing subsequent IP fragments
+	otx.AddFlow(generateBaseServiceRule(service.Spec.ClusterIP) + ", ip_frag=later" + action)
+
+	for _, port := range service.Spec.Ports {
+		baseRule, err := generateBaseAddServiceRule(service.Spec.ClusterIP, port.Protocol, int(port.Port))
+		if err != nil {
+			glog.Errorf("Error creating OVS flow for service %v, netid %d: %v", service, netID, err)
+		}
+		otx.AddFlow(baseRule + action)
+	}
+
+	return otx.EndTransaction()
+}
+
+func (oc *ovsController) DeleteServiceRules(service *kapi.Service) error {
+	otx := oc.ovs.NewTransaction()
+	otx.DeleteFlows(generateBaseServiceRule(service.Spec.ClusterIP))
+	return otx.EndTransaction()
+}
+
+func generateBaseServiceRule(IP string) string {
+	return fmt.Sprintf("table=60, ip, nw_dst=%s", IP)
```

Comment:
May be we should name the tables later (ServiceTable="table=60") so that it is easy to read and don't need to cross verify.